src/e2e-test/java/io/cdap/plugin/gcs/GCSValidationHelper.java

```diff
@@ -0,0 +1,224 @@
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * GCS Validation Helper.
+ */
+public class GCSValidationHelper {
+  private static final String avroFilePath = PluginPropertyUtils.pluginProp("gcsAvroExpectedFilePath");
+  private static final String csvFilePath = PluginPropertyUtils.pluginProp("gcsCsvExpectedFilePath");
+  private static final String projectId = PluginPropertyUtils.pluginProp("projectId");
+  private static final Gson gson = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(GCSValidationHelper.class);
+
+  /**
+   * Validates if the data in a (GCS) bucket matches the expected CSV data in JSON format.
+   *
+   * @param bucketName The name of the GCS bucket to validate.
+   * @return True if the GCS CSV data matches the expected data, false otherwise.
+   * @throws IOException If an IO error occurs during data retrieval.
+   */
+  public static boolean validateGCSSourceToGCSSinkWithCSVFormat(String bucketName) {
+    Map<String, JsonObject> expectedCSVData = readCsvAndConvertToJson();
+    Map<String, JsonObject> actualGcsCsvData = listBucketObjects(bucketName, DataFormat.CSV);
+
+    boolean isMatched = actualGcsCsvData.equals(expectedCSVData);
+
+    return isMatched;
+  }
+
+  /**
+   * Validates if the data in a (GCS) bucket matches the data
+   * obtained from Avro files converted to JSON.
+   *
+   * @param bucketName The name of the GCS bucket to validate.
+   * @return True if the GCS data matches the Avro data, false otherwise.
+   * @throws IOException If an IO error occurs during data retrieval.
+   */
+  public static boolean validateGCSSourceToGCSSinkWithAVROFormat(String bucketName) throws IOException {
+    Map<String, JsonObject> expectedAvroData = convertAvroToJsonWithKeys();
+    Map<String, JsonObject> actualGcsData = listBucketObjects(bucketName, DataFormat.JSON);
+
+    boolean isMatched = actualGcsData.equals(expectedAvroData);
+
+    return isMatched;
+  }
+
+  public static Map<String, JsonObject> listBucketObjects(String bucketName, DataFormat dataFormat) {
+    Map<String, JsonObject> bucketObjectData = new HashMap<>();
+    Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
+    Page<Blob> blobs = storage.list(bucketName);
+
+    List<Blob> bucketObjects = StreamSupport.stream(blobs.iterateAll().spliterator(), true)
+      .filter(blob -> blob.getSize() != 0)
+      .collect(Collectors.toList());
+
+    Stream<String> objectNamesWithData = bucketObjects.stream().map(blob -> blob.getName());
+    List<String> bucketObjectNames = objectNamesWithData.collect(Collectors.toList());
+
+    if (!bucketObjectNames.isEmpty()) {
+      String objectName = bucketObjectNames.get(0);
+      if (objectName.contains("part-r")) {
```

Comment:
About this part-r check here .. do you see any part-r files created in gcs bucket for your pipelines. I believe part-r files are only created for File Plugin as a source.

We can remove this check if you dont see any part-r files for your pipelines.

Reply:
We do have a part-r directory in the case for creation of a bucket in GCS Sink using a hook and as it is required to fetch and read the data output file which present in part-r.